### ISSUE TITLE
Discovery: Determine work to move collections out of GSA CKAN core and into an extension

### ISSUE THREAD

**Issue by kimwdavidson** [GSA/data.gov#1461]
### Summary 

This is a timeboxed effort to understand the technical implications of, and determine the HL level of effort involved in, moving collections out of catalog-app (CKAN 2.3) and into an extension that works with the OKF main CKAN branch (CKAN 2.8) (running in Docker). 

The technical criteria to assess the solution against will be:
* Ease of fit in the CKAN framework eco-system (will it work with the other extensions / inventory(?)
* Search will work (we can index etc)
* How elegant a solution is it (ie not incur technical debt)

Time-boxed at 3ish days:
* 0.5 to 1 day Understanding the code-base (we already have done pre-work here)
* 0.5  Summary of "non-core code" in the catalog app (ie BSP config, collections, other features - but this more for background... BSP config we can ignore... we want to focus on the collection code)
* 0.5 days High-level understanding Dept Ed collections approach using group id = collection
* 1 to 1.5 days analyzing an approach and coming up with HL solution and estimate

This discovery task is focused on purely looking at this problem from a "can we build collections as an extension" stand-point, the following are out of scope:
* Product considerations: Do we need this? What do we need? we assume we need this and we will build the same features (if this solution looks viable then we can see features collections would be required.
* "Change all WAF collections to be single file harvest, and have metadata point to resource endpoints that are relevant for download. This still doesn't handle the POD collections implementation. CKAN does have a native feature to "connect" datasets, we could consider going that route for those datasets..." ([From James Slack dated 1/31/2020](https://docs.google.com/document/d/1ZkRXFfHrlWYyTs7fQnfzNhHT4_amzyUgM5m0qrT_RXg/edit?ts=5e724f6e#)

### Links

* [Previous thoughts on Collections](https://docs.google.com/document/d/1ZkRXFfHrlWYyTs7fQnfzNhHT4_amzyUgM5m0qrT_RXg/edit?ts=5e724f6e#) 

## Acceptance Criteria
* [x] A proposed solution to move the Collections feature out of CKAN core and have a technical understanding of the implications of doing that on the other extensions
* [x] Estimate work for a solution (or a spike) to move collections out of CKAN core and into a separate extension 
* [ ] Estimate how long it would take to amend the extensions that have a reverse dependency with the collections code (Basically how long would it take to change the extensions so they work with a new collections extension)

## Tasklist - getting set up for this task
- [x] Esteban to draft the issue and discuss with Julie
- [x] Julie to ask Tom to add you to Dept Ed repo (so you can see code (this project is temporarily no longer open source  :angry:)
- [x] Esteban to provide link to documentation and point Julie to the Dept Ed code and answer any questions she has
- [x] Understand collections implementation in UI
    - [x] Esteban to do first pass
    - [x] Julie to review document and add/amend

## Task list - analysis
* [ ] Understanding the current approach 
    - [x] What metadata to they have associated with Collections, what are the features of collections?
    - [ ] Understand the dependencies needed by collections and reverse dependencies (extensions use collections code)
* [ ] Run the OKF upstream locally to better understand how features the collections are coming from the extensions or core, and where the collections code that exists in CKAN core is being used. 
* [ ] If we move this collections code out of core and into an extension - what do we break? 
* [x] Create a list of possible technical ideas and provide pros and cons (Esteban's initial ideas of our options: keep it in the core (boo), CKAN ext scheming (I'd stay clear of that), use group) 

### Recommendation

There are 2 options to deal with collections:
1. Re-implement the elegant hack in CKAN 2.8 (fork again) - Not recommended
2. **Create an extension - Recommended**

If we opt for an extension there are a number of options. 
1. Implement the package id = collections as an extension. 
   *  This will work fine for the existing collections features.
2. Use group id = collections as per Dept Ed solution
   * This is a good option if and only if you want to have more features. This would involve rework for the other extensions in addition to the work involved in creating said features 
3. Use Scheming 
   * Datopian evaluated this when building the collections feature for Dept Ed and decided against using scheming. To be honest the scheming extension has a mixed reception in the CKAN community. Some like it more than others but the main issue working with scheming is with scheming "changing something here breaks something that is unrelated over there" (which some might say is a general CKAN issue) so this is not recommended. 

Therefore the recommendation is to build an extension using the package id = collection pattern. If a richer set of features were required in the future then the extension can be upgraded, possibly using the group id = collections pattern. 

As a t-shirt size, we would estimate this to be 2 to 3 weeks of work of which half that time is developing unit tests. But we would know after 2/3 days of dev time whether this is going to work. 

Proposed next steps:
* Understand the other code in catalog to make sure that we incorporate that as well. See [the sister issue](https://app.zenhub.com/workspaces/datagov-multi-tenant-ckan-5cf12b601437ff1e943f0aa2/issues/gsa/datagov-ckan-multi/257).
* Create a ticket to build a collection extension
   * Estimate the task (ie move beyond a high-level estimation provided as part discovery task)
   * Create extension
   * Validate that dependencies and reverse dependencies are still working 
   * Create unit tests
   * Created another issue to make sure that the way other harvester types are considered in addition to data.json harvester and DCAT-US schema which use the ispartof. That is to say, we need to make sure that our proposed solution works for all harvesters. 
   * Note to team - let's not forget to DevSecOps considerations - need to rope Tom in early

### Analysis

[See more details in our scratchpad analysis doc](https://hackmd.io/DP3btMgNQQ6xoxuRlM2Ajw)

### Previous issues related to this

[CKAN 2.8 plan of work](https://github.com/GSA/datagov-deploy/wiki/Catalog-CKAN-2.8-upgrade)

### Where to start looking for the Collections code 

Part of the data.json standard defined in the Project Open Data is [the field isPartOf](https://project-open-data.cio.gov/v1.1/schema/#isPartOf). This field allows the grouping of multiple datasets into a “collection”. This field should be employed by the individual datasets that together make up a collection. The value for this field should match the identifier of the parent dataset.  

GSA added this functionality to CKAN. In packages, GSA used some new `extras`:
 - _collection_package_id_: Refers to the package_id of the _father_.
 - _is_collection_: True for the packages that have _children_.
 
Both seem to have been  [included in CKAN searches in 2015](https://github.com/avdata99/ckan/commit/3ef9f1be9667476d14ce4411d0e36bbdf81e00dc). Also in 2015 the [added to datajson ext](https://github.com/GSA/ckanext-datajson/commit/6a44b6896e8cb73a61f13dbed667b46d183bcffb#diff-171e908c6fdcbd1f40bb8081a6a15ac2R649). Also [added the is_collection](https://github.com/GSA/ckanext-datajson/commit/6a44b6896e8cb73a61f13dbed667b46d183bcffb#diff-171e908c6fdcbd1f40bb8081a6a15ac2R296) extra. GSA also uses this _collection_package_id_ [at the DataGovTheme](https://github.com/GSA/ckanext-datagovtheme/commit/051a2a9bf5ba624332ad7444e0946823698f235f#diff-7dcc34d0dd09e7870eb542c3b8ff914fR36).  

From Aaron about the where to find the collections code:
> "One gotcha, isPartOf is specific to the data.json harvester and DCAT-US schema but other harvesters might implement collections too e.g. WAF or CSW. It's unclear how many harvesters have implemented it and its unclear if any harvest sources are actually using it."

### Some info on how collections are implemented in the current UI

In summary, the features of the data.gov 
* If a dataset is part of a collection you can go to the collection
* Search within a collection
* Not easily discoverable

[See this document for more info](https://hackmd.io/aqq0bKERSIC6h4Bbj6-Tiw?both)

**Comment by estebanruseler** [comment]
@thejuliekramer as promised, here's the [details of how the collections are done in the current UI](https://hackmd.io/aqq0bKERSIC6h4Bbj6-Tiw) based on what I found. But please double-check this...

**Comment by estebanruseler** [comment]
Julie and my wip analysis https://hackmd.io/DP3btMgNQQ6xoxuRlM2Ajw

**Comment by estebanruseler** [comment]
@thejuliekramer as promised I’ve updated the analysis doc with eg.gov collections features and code snippets.

**Comment by thejuliekramer** [comment]
@estebanruseler  After poking around for a few days it seems like we can move the [methods](https://github.com/GSA/ckan/commit/3ef9f1be9667476d14ce4411d0e36bbdf81e00dc) we've changed for the collections search into an extension and override the existing methods in ckan... the same methods still exist in newer versions of CKAN... I think they will interact the same with the code the exists in datajson and geodatagov extensions as far as I can tell, but I will know more if we get approval to do a spike, I ran out of time to actually fully get it working locally. I would estimate the work to move the code into an extension would take about a week - and testing the feature would take another week since no tests exist. This is also all assuming we want to keep the same exact functionality and UI we have currently and not add anything else (like they did in Dept of Ed).  Let me know your thoughts

**Comment by mogul** [comment]
@jbrown-xentity FYI! @estebanruseler is available to talk about the findings in more detail if you like, especially if it's useful to you in the context of other projects.

**Comment by jbrown-xentity** [comment]
@estebanruseler would love to sync up about your findings and next steps for this!

**Comment by estebanruseler** [comment]
@jbrown-xentity @thejuliekramer and I were working on this together so we should have a chat the 3 of us. Tomorrow my day is pretty full but should we meet for 30 mins on Monday. I'm available from 1.30 to 3 pm ET.

**Comment by adborden** [comment]
I was under the impression that it was not easy to patch/extend CKAN's search functionality from an extension, but if that's not the case then the approach sounds great.
